File: test/models.test.ts
import { test } from 'node:test';
import assert from 'node:assert/strict';
import type { ModelInfo } from '../src/core/model.ts';
import { ALL_EFFORTS, effortOptions, modelOptions, selectedModel } from '../src/core/models.ts';

// What Claude Code 2.1.280 answered to initialize on a Team account, trimmed.
const models: ModelInfo[] = [
  { value: 'default', resolvedModel: 'claude-opus-5-5[1m]', displayName: 'Default (recommended)', description: 'Opus 5.5 with 1M context', supportedEffortLevels: ['low', 'medium', 'high', 'xhigh', 'max'] },
  { value: 'opus[1m]', resolvedModel: 'claude-opus-5-5[1m]', displayName: 'Opus (1M context)', supportedEffortLevels: ['low', 'medium', 'high', 'xhigh', 'max'] },
  { value: 'claude-fable-5-1[1m]', resolvedModel: 'claude-fable-5-1', displayName: 'Fable', supportedEffortLevels: ['low', 'medium', 'high', 'xhigh', 'max'] },
  { value: 'sonnet', resolvedModel: 'claude-sonnet-5', displayName: 'Sonnet', supportedEffortLevels: ['low', 'medium', 'high'] },
  { value: 'haiku', resolvedModel: 'claude-haiku-4-5-20251001', displayName: 'Haiku' },
];

test('modelOptions follows the harness list, default row becomes the empty value', () => {
  assert.deepEqual(
    modelOptions(models).map(([v, l]) => [v, l]),
    [
      ['', 'Default (recommended)'],
      ['opus[1m]', 'Opus (1M context)'],
      ['claude-fable-5-1[1m]', 'Fable'],
      ['sonnet', 'Sonnet'],
      ['haiku', 'Haiku'],
    ],
  );
  assert.equal(modelOptions(models)[0]?.[2], 'Opus 5.5 with 1M context');
  // Before the harness has answered there is nothing to list but the default.
  assert.deepEqual(modelOptions(undefined), [['', 'Default model']]);
  assert.deepEqual(modelOptions([]), [['', 'Default model']]);
});

test('selectedModel maps what the session reports onto a picker row', () => {
  assert.equal(selectedModel(models, ''), '');
  assert.equal(selectedModel(models, 'sonnet'), 'sonnet'); // what we asked for via set_model
  assert.equal(selectedModel(models, 'default'), '');
  assert.equal(selectedModel(models, 'claude-opus-5-5[1m]'), ''); // init reported the resolved default
  assert.equal(selectedModel(models, 'claude-opus-5-5'), ''); // transcript drops the context suffix
  assert.equal(selectedModel(models, 'claude-fable-5-1'), 'claude-fable-5-1[1m]');
  assert.equal(selectedModel(models, 'claude-mystery-9'), 'claude-mystery-9'); // unknown: keep so the picker shows it
  assert.equal(selectedModel(undefined, 'sonnet'), 'sonnet');
});

test('effortOptions narrows to what the selected model supports', () => {
  assert.deepEqual(effortOptions(models, 'sonnet').map(([v]) => v), ['', 'low', 'medium', 'high']);
  assert.deepEqual(effortOptions(models, '').map(([v]) => v), ['', 'low', 'medium', 'high', 'xhigh', 'max']);
  assert.deepEqual(effortOptions(models, 'haiku'), [['', 'Default effort']]);
  assert.equal(effortOptions(models, 'claude-mystery-9'), ALL_EFFORTS);
  assert.equal(effortOptions(undefined, 'sonnet'), ALL_EFFORTS);
});

File: daemon/internal/managed/models_test.go
package managed

import (
	"encoding/json"
	"testing"
)

// A trimmed copy of what Claude Code 2.1.280 answers to an initialize control_request.
const initResponse = `{"commands":[],"models":[
 {"value":"default","resolvedModel":"claude-opus-5-5[1m]","displayName":"Default (recommended)","description":"Opus 5.5 with 1M context","supportsEffort":true,"supportedEffortLevels":["low","medium","high","xhigh","max"],"supportsFastMode":true},
 {"value":"opus[1m]","resolvedModel":"claude-opus-5-5[1m]","displayName":"Opus (1M context)","description":"Opus 5.5 with 1M context","supportsEffort":true,"supportedEffortLevels":["low","medium","high","xhigh","max"]},
 {"value":"sonnet","resolvedModel":"claude-sonnet-5","displayName":"Sonnet","description":"Sonnet 5","supportsEffort":true,"supportedEffortLevels":["low","medium","high"],"disabled":true},
 {"value":"haiku","resolvedModel":"claude-haiku-4-5-20251001","displayName":"Haiku","description":"Haiku 4.5"}
],"unavailable_models":[],"account":{"email":"x@y"}}`

func TestParseModels(t *testing.T) {
	got := parseModels(json.RawMessage(initResponse))
	if len(got) != 3 {
		t.Fatalf("want 3 rows (disabled dropped), got %d: %+v", len(got), got)
	}
	if got[0].Value != "default" || got[0].ResolvedModel != "claude-opus-5-5[1m]" || got[0].DisplayName != "Default (recommended)" {
		t.Errorf("first row: %+v", got[0])
	}
	if len(got[0].SupportedEffortLevels) != 5 || got[0].SupportedEffortLevels[4] != "max" {
		t.Errorf("effort levels: %v", got[0].SupportedEffortLevels)
	}
	if got[1].Value != "opus[1m]" || got[2].Value != "haiku" {
		t.Errorf("order/filter: %s, %s", got[1].Value, got[2].Value)
	}
	if got[2].SupportedEffortLevels != nil {
		t.Errorf("haiku should take no effort setting, got %v", got[2].SupportedEffortLevels)
	}
	if out, _ := json.Marshal(got[2]); string(out) != `{"value":"haiku","resolvedModel":"claude-haiku-4-5-20251001","displayName":"Haiku","description":"Haiku 4.5"}` {
		t.Errorf("wire shape: %s", out)
	}
}

func TestParseModelsGarbage(t *testing.T) {
	if got := parseModels(json.RawMessage(`not json`)); got != nil {
		t.Errorf("want nil, got %v", got)
	}
	if got := parseModels(json.RawMessage(`{"models":[]}`)); len(got) != 0 {
		t.Errorf("want empty, got %v", got)
	}
}
